packages/vhd-lib/src/Vhd/VhdDirectory.js

```diff
@@ -0,0 +1,186 @@
+    assert(Buffer.isBuffer(buffer))
+    // here we can implement compression and / or crypto
+    return this._handler.writeFile(this.getChunkPath(partName), buffer, { flags: this._flags })
+  }
+
+  async readHeaderAndFooter() {
+    const { buffer: bufHeader } = await this._readChunk('header')
+    const { buffer: bufFooter } = await this._readChunk('footer')
+    const footer = buildFooter(bufFooter)
+    const header = buildHeader(bufHeader, footer)
+
+    this.footer = footer
+    this.header = header
+  }
+
+  async readBlock(blockId, onlyBitmap = false) {
+    if (onlyBitmap) {
+      throw new Error(`reading 'bitmap of block' ${blockId} in a VhdDIrecotry is not implemented`)
+    }
+    const { buffer } = await this._readChunk(blockId)
+    return {
+      id: blockId,
+      bitmap: buffer.slice(0, this.bitmapSize),
+      data: buffer.slice(this.bitmapSize),
+      buffer,
+    }
+  }
+  ensureBatSize() {
+    // nothing to do in directory mode
+  }
+
+  async writeFooter() {
+    const { footer } = this
+
+    const rawFooter = fuFooter.pack(footer)
+
+    footer.checksum = checksumStruct(rawFooter, fuFooter)
+    debug(`Write footer  (checksum=${footer.checksum}). (data=${rawFooter.toString('hex')})`)
+
+    await this._writeChunk('footer', rawFooter)
+  }
+
+  writeHeader() {
+    const { header } = this
+    const rawHeader = fuHeader.pack(header)
+    header.checksum = checksumStruct(rawHeader, fuHeader)
+    debug(`Write header  (checksum=${header.checksum}). (data=${rawHeader.toString('hex')})`)
+    return this._writeChunk('header', rawHeader)
+  }
+
+  writeBlockAllocationTable() {
+    assert.notStrictEqual(this.blockTable, undefined, 'Block allocation table has not been read')
+    assert(this.blockTable.length, 'Block allocation table is empty')
+    assert(this.blockTable.length % 4 === 0, 'Block allocation table size is incorrect')
+
+    const { blockTable } = this
+
+    /**
+     * In directory mode, the bat is a sequence of bits
+     * A zero bit indicates that the block is not present
+     */
+    const buffer = Buffer.alloc(this.header.maxTableEntries, 0)
+    for (let blockId = 0; blockId < blockTable.length / 4; blockId++) {
+      if (this.blockTable.readUInt32BE(blockId * 4) !== BLOCK_UNUSED) {
+        setBitmap(buffer, blockId)
+      }
+    }
+    return this._writeChunk('bat', buffer)
+  }
+
+  setUniqueParentLocator(fileNameString) {
+    const encodedFilename = Buffer.from(fileNameString, 'utf16le')
+    return this._write('parentLocator0', encodedFilename)
+  }
+
+  // only works if data are in the same bucket
+  // and if the full block is modified in child ( which is the case whit xcp)
+
+  coalesceBlock(child, blockId) {
+    this._handler.copy(child.getChunkPath(blockId), this.getChunkPath(blockId))
+  }
+
+  async writeEntireBlock(block) {
+    // @todo should check if bat should be updated
+    await this._writeChunk(block.id, block.buffer)
+  }
+
+  readParentLocatorData(parentLocatorId) {
```

Comment:
If you want, you can move this method to the parent class, do all the checks there, and only delegate the true read to a simple `_readParentLocatorData`.